### PR TITLE
Replace spirv_std::storage_class::X<T> with T/&mut T and #[spirv(x)] &T/&mut T.

### PR DESCRIPTION
**Pull request by eddyb**
~~This is my pre-#416 branch, I started this as part of #414, but only later realized I don't need it.~~

~~It's easier than I expected because, once again, I was able to mostly reuse existing logic (attribute parsing in this case).~~

As it's currently implemented, rules are:
* only `&T` and `&mut T` are allowed as entry parameter types, with explicit storage class attributes
  * should probably be `&mut MaybeUninitialized<T>` instead of `&mut T`, but ideally we can use `-> T` instead for `Output`
  * it considers `&T` immutable but it should also check that `T: Freeze`
* `T` defaults to `Input`, `&mut T` defaults to `Output`, and those storage classes cannot be explicitly specified
* all other storage classes use a `#[spirv(...)]` attribute on the parameter, e.g. this line in `sky-shader`:
  ```rust
      #[spirv(push_constant)] constants: &ShaderConstants,
  ```
* `UniformConstant`, and `PushContant` storage classes only allow `&T` (i.e. not `&mut T`) types
* `Private`, `Function` and `Generic` storage classes cannot be specified at all
  * should probably restrict this further based on shader stage or kernel vs shader, but we can do that later

~~This works and passes tests, but I'm opening it as a draft because it should wait on a decision regarding #416, especially around the `Output` storage class and whether we should use a `->` style, support bundling them in `struct`s etc.~~

**EDIT**: updated to reflect PR changes, i.e. `Input` not taking references anymore.

### CODE REVIEW

**Review by khyperia — APPROVED** [review]
Looks hecking great! @hrydgard and I had a conversation in discord, path forward seems to be to make inputs be by-value, and then in a follow-up, do the annoyingly complex work of making outputs be returned via tuple or struct or whatever.

**Comment by Hentropy** [comment]
I mentioned this on the MCP as well, but representing PhysicalStorageBuffer as &/&mut would be really neat so switching to by-value for Input/Output would be wonderful.

**Comment by eddyb** [comment]
`mut my_output: f32` just means `my_output: f32` with a `let mut my_output = my_output;` in the body.

*Even if* Rust-GPU wanted to do something funky, we don't *really* have control over what things like that mean.
What happens today is the SPIR-V entry-point calls the *regular* Rust `#[spirv(entry)] fn foo(...) {...}` with the right arguments.

**Comment by eddyb** [comment]
@khyperia FWIW this is ready for a re-review, but I can't figure out what's wrong with the CI.
I don't want to tell it to retry because it might auto-merge because of the previous review.

**Comment by XAMPPRocky** [comment]
I'm going to re-run and temporarily disable auto-merge

**Comment by XAMPPRocky** [comment]
Yeah, it was an issue with GitHub Actions, the PR passes.

File: crates/rustc_codegen_spirv/src/attr.rs [hunk 1]
```diff
@@ -298,6 +297,31 @@ impl CheckSpirvAttrVisitor<'_> {
                                 span,
                                 "attribute is only valid on a parameter of an entry-point function",
                             );
+                        } else {
+                            // FIXME(eddyb) should we just remove all 5 of these storage class
+                            // attributes, instead of disallowing them here?
```
Comment (XAMPPRocky) [inline crates/rustc_codegen_spirv/src/attr.rs:302]: Would there ever be a need to use these attributes in `spirv-std`? If not, I agree that we should remove them.

**Review by XAMPPRocky — APPROVED** [review]

**Review by khyperia — APPROVED** [review]

**Pull request merged**